### DISCUSSION TITLE
Generate DataClass with joined values

### DISCUSSION THREAD

**Discussion by II11II** [simolus3/drift#1884] [Q&A] [answered]
I am newbie in drift. 
Is it possible to generate DataClass with Joined Values by reference key?
Currently, the drift generates DataClass with reference workPlaceId 
**Actual Result:**
```
class Doctor extends DataClass implements Insertable<Doctor> {
  final String firstName;
  final String lastName;
  final String? avatarPath;
  final Gender? gender;
  final bool isOnline;
  final int id;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int? workPlaceId;  // generate only id 
  final double? height;
  final double? weight;
  final DateTime? startDateJob;
  ```
 **Expected result:**
 
  ```
class Doctor extends DataClass implements Insertable<Doctor> {
  final String firstName;
  final String lastName;
  final String? avatarPath;
  final Gender? gender;
  final bool isOnline;
  final int id;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final WorkPlace workPlace; // generate class instead of its id
  final double? height;
  final double? weight;
  final DateTime? startDateJob;
  ```

**Comment by simolus3** [comment, marked as answer]
Drift does not generate classes with joined values for you. The reason is that, for the most part, drift's Dart API is a close wrapper around SQL. If we had those implicit joins in the data class, drift would also have to rewrite your queries at runtime to add that implicit join and load the values.

If you have a specific SQL query (with a join) and you want drift to generate a class _for that query_, you can do that! Drift can [analyze your SQL queries](https://drift.simonbinder.eu/docs/using-sql/) at build time, figure out what columns they return and write a data class based on that. So you could add something like this to your database class for instance:

```dart
@DriftDatabase(
  tables: [Doctors, Workplaces],
  queries: {
    'doctorsWithWorkplace':
        'SELECT *, workplace.** FROM doctors LEFT JOIN workplaces workplace ON workplace.id = doctors.work_place'
  },
)
```

The `.**` will cause drift to [nest](https://drift.simonbinder.eu/docs/using-sql/drift_files/#nested-results) the workplace columns into the main class which should make it generate the class you had in mind.

**Reply by II11II** [reply]
P.s. https://github.com/simolus3/drift/discussions/1884#discussioncomment-2936074

**Comment by II11II** [comment]
@simolus3 I tried your suggestion but got the class which has attributes in it from class **Doctors** and **Workplaces**

```
class GetDoctorByIdResult {
  final String firstName;
  final String lastName;
  final String? avatarPath;
  final Gender? gender;
  final bool isOnline;
  final int id;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int? workPlaceId;
  final double? height;
  final double? weight;
  final DateTime? startDateJob;
  final int? id1;
  final DateTime? createdAt1;
  final DateTime? updatedAt1;
  final String? imagePath;
  final String? localeId;
  final int? locationId;
  GetDoctorByIdResult({
    required this.firstName,
    required this.lastName,
    this.avatarPath,
    this.gender,
    required this.isOnline,
    required this.id,
    required this.createdAt,
    this.updatedAt,
    this.workPlaceId,
    this.height,
    this.weight,
    this.startDateJob,
    this.id1,
    this.createdAt1,
    this.updatedAt1,
    this.imagePath,
    this.localeId,
    this.locationId,
  });
}
```
that does not satisfy the expected result:

```
class GetDoctorByIdResult {
  final String firstName;
  final String lastName;
  final String? avatarPath;
  final Gender? gender;
  final bool isOnline;
  final int id;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int? workPlaceId;
  final double? height;
  final double? weight;
  final WorkPlace? workPlace;
  GetDoctorByIdResult({
    required this.firstName,
    required this.lastName,
    this.avatarPath,
    this.gender,
    required this.isOnline,
    required this.id,
    required this.createdAt,
    this.updatedAt,
    this.workPlaceId,
    this.height,
    this.weight,
    this.workPlace
  });
}
```

**Reply by simolus3** [reply]
Did you write the query with two stars for the nested result? I've tried it again with a similar schema structure from a drift example and it worked fine :thinking: Do you have a repro that you could share (say the table's definitions and the query you added)?

**Reply by II11II** [reply]
Sorry, I am not able to share the whole project. 
I tried to create a gist with the dependent classes and left comments. 
[Gist link](https://gist.github.com/II11II/41a0d0f05a3fe997d944a2257f25f34b)

**Reply by II11II** [reply]
Btw, do we have a group(f.e. in discord) to discuss such kinds of questions?

**Reply by North101** [reply]
> Sorry, I am not able to share the whole project. I tried to create a gist with the dependent classes and left comments. [Gist link](https://gist.github.com/II11II/41a0d0f05a3fe997d944a2257f25f34b)

I think the SQL should be:
```
SELECT doctor_table.*, work_place_table.** FROM doctor_table LEFT JOIN work_place_table work_place ON work_place.id = doctor_table.work_place_id
```

**Reply by II11II** [reply]
Hi @North101 
I tried this query and got this result:

```

mixin _$DoctorRepositoryMixin on DatabaseAccessor<AppDatabase> {
  $DoctorTableTable get doctorTable => attachedDatabase.doctorTable;
  $WorkPlaceTableTable get workPlaceTable => attachedDatabase.workPlaceTable;
  Selectable<Doctor> getDoctor() {
    return customSelect(
        'SELECT doctor_table.*, work_place_table.** FROM doctor_table LEFT JOIN work_place_table AS work_place ON work_place.id = doctor_table.work_place_id',
        variables: [],
        readsFrom: {
          doctorTable,
          workPlaceTable,
        }).map(doctorTable.mapFromRow);
  }
}

``` 

```
class Doctor extends DataClass implements Insertable<Doctor> {
  final String firstName;
  final String lastName;
  final String? avatarPath;
  final Gender? gender;
  final bool isOnline;
  final int id;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int? workPlaceId; // it is id but not reference to WorkPlace class
  final DateTime? startDateJob;
```

**Reply by North101** [reply]
> Hi @North101 I tried this query and got this result:
> 
> ```
> 
> mixin _$DoctorRepositoryMixin on DatabaseAccessor<AppDatabase> {
>   $DoctorTableTable get doctorTable => attachedDatabase.doctorTable;
>   $WorkPlaceTableTable get workPlaceTable => attachedDatabase.workPlaceTable;
>   Selectable<Doctor> getDoctor() {
>     return customSelect(
>         'SELECT doctor_table.*, work_place_table.** FROM doctor_table LEFT JOIN work_place_table AS work_place ON work_place.id = doctor_table.work_place_id',
>         variables: [],
>         readsFrom: {
>           doctorTable,
>           workPlaceTable,
>         }).map(doctorTable.mapFromRow);
>   }
> }
> ```
> 
> ```
> class Doctor extends DataClass implements Insertable<Doctor> {
>   final String firstName;
>   final String lastName;
>   final String? avatarPath;
>   final Gender? gender;
>   final bool isOnline;
>   final int id;
>   final DateTime createdAt;
>   final DateTime? updatedAt;
>   final int? workPlaceId; // it is id but not reference to WorkPlace class
>   final DateTime? startDateJob;
> ```

Did you rerun the buildrunner?
```
flutter pub run buildrunner build
```

**Reply by II11II** [reply]
yes

**Reply by North101** [reply]
Hmm. I usually use .drift files. Maybe there's an issue with `queries`?

query.drift
```sql
doctorsWithWorkplace:
SELECT
  doctor_table.*,
  work_place_table.**
FROM doctor_table
LEFT JOIN work_place_table work_place
  ON work_place.id = doctor_table.work_place_id
```

db.dart
```dart
@DriftDatabase(
  // relative import for the drift file. Drift also supports `package:`
  // imports
  include: {'query.drift'},
)
class AppDb extends _$AppDb {
  ...
}
```

**Reply by simolus3** [reply]
Your query generates these warnings for me:

```
The sql query getDoctor is invalid: line 1, column 11: Unknown table: work_place_table
  ╷
1 │ SELECT *, work_place_table.** FROM doctor_table LEFT JOIN work_place_table work_place ON work_place.id = doctor_table.work_place_id
  │           ^^^^^^^^^^^^^^^^^^^
  ╵
[INFO] drift_dev:drift_dev on example/repro.dart:Lint for getDoctor: line 1, column 11: Nested star columns must refer to a table directly. They can't refer to a table-valued function or another select statement.
  ╷
1 │ SELECT *, work_place_table.** FROM doctor_table LEFT JOIN work_place_table work_place ON work_place.id = doctor_table.work_place_id
  │           ^^^^^^^^^^^^^^^^^^^
  ╵
```

And sure enough, since you joined the table as `JOIN work_place_table work_place`, that's the name you need to use to refer to that table (just like you would in SQL). For me, it works after I change the SQL to `SELECT *, work_place.** FROM ...` (reference: https://gist.github.com/simolus3/127cae2672e5f3e115b8699bdcb199ff). It generates a class like this:

```dart
class GetDoctorResult extends CustomResultSet {
  final String firstName;
  // ...
  final int? locationId;
  final WorkPlace? workPlace; // !!

```

Now I'm wondering if you haven't received that error?

**Reply by II11II** [reply]
@simolus3 I tried this query and in fact, I got a nested one. However, there are fields created from **WorkPlace** class inside **GetDoctorResult** class. 
Is it okay if **GetDoctorResult** is not extending  from **CustomResultSet** while i am generating it? 
```
class GetDoctorResult {
  final String firstName;
  final String lastName;
  final String? avatarPath;
  final Gender? gender;
  final bool isOnline;
  final int id;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int? workPlaceId;
  final DateTime? startDateJob;
  final int? id1; // bad, i can access through workPlace.id
  final DateTime? createdAt1; // bad, i can access through workPlace.createdAt1
  final DateTime? updatedAt1; // bad, i can access through workPlace.updatedAt1
  final String? imagePath; // bad, i can access through workPlace.imagePath
  final String? localeId; // bad, i can access through workPlace.localeId
  final int? locationId; // bad, i can access through workPlace.locationId
  final WorkPlace? workPlace; // Good 
```

**Reply by simolus3** [reply]
Sorry for the back and forth here. This is working as intended, I just gave you a bad query: The `*` in `SELECT *, work_place.**` resolves to all columns from the query, which includes joined tables. So drift_dev is generating what  it should be generating, it's just not helpful.

Those columns should go away when doing `SELECT doctor_table.*, work_place.**`.

**Reply by II11II** [reply]
It is working now🙂

**Reply by II11II** [reply]
@simolus3 How to generate a class with a custom name while generating entities through a query?
For now, I am getting class name as ```GetDoctorResult``` but it would more meaningful if the generator generated me  ```DoctorEntity```

**Reply by simolus3** [reply]
> How to generate a class with a custom name while generating entities through a query?

This is only possible with [drift files](https://drift.simonbinder.eu/docs/using-sql/drift_files/#result-class-names).

**Reply by II11II** [reply]
@simolus3 thank you
